### PR TITLE
Add siteSlug prop to CompositeCheckoutContainer

### PR DESCRIPTION
**Pull request by nbloomf**
#### Changes proposed in this Pull Request

* All control paths into the `checkout` middleware, where new checkout will be rendered, will have already created a cart on the server. This means that to get the canonical cart data that we need it's enough to call `getCart` with the cart slug.

We also specialize the `setCart` handler to fix the key parameter.

To do:
- [x] Rebase after #37977
- [x] Merge in Filippo's changes from #37941
- [x] Get removeItem to work

#### Testing instructions
Automated tests: `npm run test-packages composite-checkout-wpcom`

Manual test:
- Run calypso with `ENABLE_FEATURES=composite-checkout-wpcom npm start`
- Go through checkout flow for a new site. The selected products should load in the checkout form.

### CODE REVIEW

**Comment by sirbrillig** [comment]
Great work identifying the types of carts and typing this code!

**Comment by sirbrillig** [comment]
This will probably have a couple of conflicts with #37937 but I think they're relatively minor.

File: client/my-sites/checkout/checkout/composite-checkout-container.js [hunk 1]
```diff
@@ -19,52 +19,19 @@ import {
...
 
 const wpcom = wp.undocumented();
 
-const useShoppingCart = makeShoppingCartHook(
-	( cartKey, cartParam ) => wpcom.setCart( cartKey, cartParam ),
-	initialCart
-);
+export const getServerCart = wpcom.getCart;
+
+const useShoppingCart = cartKey => {
```
Comment (sirbrillig) [inline client/my-sites/checkout/checkout/composite-checkout-container.js]: I think it might be best not to start this function with `use` since it's not a hook, it's a hook creator creator. Maybe something like `buildMakeShoppingCartHook`? 🤔 Although that's pretty repetative. Is it possible to pass `cartKey` as an argument to `makeShoppingCartHook` directly?

This also makes me wonder if it's safe to be re-creating a hook on every render. With the old version of `makeShoppingCartHook`, it was called when the file loaded so there was only ever one version of the actual hook. This version will call the creator creator on every render, which will create a new version of the hook. If the hook has any internal state, I think this might cause problems.

So a better question might be: since `cartKey` is dynamic and changes on render (and is therefore a dependency for any effects of the hook), is it possible to pass `cartKey` as an argument to the hook itself, rather than preparing it here?

File: packages/composite-checkout-wpcom/src/components/wp-checkout-order-review.js [hunk 2]
```diff
@@ -18,7 +18,7 @@ import {
 	WPOrderReviewSection,
 } from './wp-order-review-line-items';
 
-export default function WPCheckoutOrderReview( { className } ) {
+export function WPCheckoutOrderReview( { className, removeItem } ) {
```
Comment (sirbrillig) [inline packages/composite-checkout-wpcom/src/components/wp-checkout-order-review.js]: I don't mind, but is there a reason to change this to a named export rather than the default? I don't see other exports...

File: packages/composite-checkout-wpcom/src/components/wp-order-review-line-items.js [hunk 3]
```diff
@@ -32,12 +31,12 @@ const OrderReviewSectionArea = styled.div`
...
 `;
 
-function WPLineItem( { item, className, hasDeleteButtons, removeProduct } ) {
+function WPLineItem( { item, className, hasDeleteButtons, removeItem } ) {
 	const translate = useTranslate();
-	const hasDomainsInCart = useHasDomainsInCart();
 	const itemSpanId = `checkout-line-item-${ item.id }`;
 	const deleteButtonId = `checkout-delete-button-${ item.id }`;
 	const [ isModalVisible, setIsModalVisible ] = useState( false );
+	const hasDomainsInCart = useHasDomainsInCart();
```
Comment (sirbrillig) [inline packages/composite-checkout-wpcom/src/components/wp-order-review-line-items.js]: Super minor, but is there any reason for moving this down, or was it just an artifact of refactoring? (In a big PR it's nice to avoid any unnecessary changes to make it easier to review.)

File: packages/composite-checkout-wpcom/src/types.ts [hunk 4]
```diff
@@ -6,24 +6,39 @@ import {
...
 	translateWpcomPaymentMethodToCheckoutPaymentMethod,
-} from './types/wpcom-payment-method-class';
-import { ServerCart, ServerCartItem } from './types/server-cart';
+} from './types/backend/payment-method';
+import {
+	RequestCart,
+	RequestCartProduct,
+	ResponseCart,
+	ResponseCartProduct,
+    emptyResponseCart,
```
Comment (sirbrillig) [inline packages/composite-checkout-wpcom/src/types.ts]: Super minor: indent is different on this line.

File: packages/composite-checkout-wpcom/src/types.ts [hunk 5]
```diff
@@ -6,24 +6,39 @@ import {
...
 	WPCOMPaymentMethodClass,
 	readWPCOMPaymentMethodClass,
 	translateWpcomPaymentMethodToCheckoutPaymentMethod,
-	ServerCart,
-	ServerCartItem,
+	RequestCart,
+	RequestCartProduct,
+	ResponseCart,
+	ResponseCartProduct,
+    emptyResponseCart,
```
Comment (sirbrillig) [inline packages/composite-checkout-wpcom/src/types.ts]: Different indent again.

File: packages/composite-checkout-wpcom/src/components/wp-order-review-line-items.js (hunk 3 above)
Comment (nbloomf) [reply packages/composite-checkout-wpcom/src/components/wp-order-review-line-items.js]: It's an artifact- thanks for the heads up.

File: client/my-sites/checkout/checkout/composite-checkout-container.js [hunk 6]
```diff
@@ -125,7 +79,10 @@ export function isApplePayAvailable() {
 
 const availablePaymentMethods = [ applePayMethod, stripeMethod, paypalMethod ].filter( Boolean );
 
-export default function CompositeCheckoutContainer() {
+const getCart = ( ...args ) => wpcom.getCart( ...args );
```
Comment (sirbrillig) [inline client/my-sites/checkout/checkout/composite-checkout-container.js]: I know I was the one to add these, but we should check to see if this is required or if we can directly pass `wpcom.getCart`/`wpcom.setCart` to the props in `WPCheckoutWrapper`. If we do need them, as I suspect, we should add a comment above this line to say something like "wpcom functions require binding to the wpcom object to function correctly".

File: client/my-sites/checkout/controller.jsx [hunk 7]
```diff
@@ -22,7 +22,7 @@ import { isGSuiteRestricted } from 'lib/gsuite';
 import { getRememberedCoupon } from 'lib/cart/actions';
 import { sites } from 'my-sites/controller';
 import config from 'config';
-import CompositeCheckoutContainer from './checkout/composite-checkout-container';
+import { CompositeCheckoutContainer } from './checkout/composite-checkout-container';
```
Comment (sirbrillig) [inline client/my-sites/checkout/controller.jsx]: `CompositeCheckoutContainer` is the default export of its file, so I don't think this should be working 🤔

File: packages/composite-checkout-wpcom/src/index.js [hunk 8]
```diff
@@ -2,9 +2,15 @@
...
-import { mockCartEndpoint } from './mock/cart-endpoint';
+import { makeShoppingCartHook } from './hooks/shopping-cart-hook';
+import { mockSetCartEndpoint, mockGetCartEndpointWith } from './mock/cart-endpoint';
 import { mockPayPalExpressRequest } from './mock/paypal-payment-method';
 
 // Re-export the public API
-export { WPCheckoutWrapper, makeShoppingCartHook, mockCartEndpoint, mockPayPalExpressRequest };
+export {
+	WPCheckoutWrapper,
+	makeShoppingCartHook,
```
Comment (sirbrillig) [inline packages/composite-checkout-wpcom/src/index.js]: I think this no longer exists and is now `useShoppingCart`

**Comment by sirbrillig** [comment]
This also needs testing instructions.

File: client/my-sites/checkout/checkout/composite-checkout-container.js (hunk 6 above)
Comment (nbloomf) [reply client/my-sites/checkout/checkout/composite-checkout-container.js]: I think this is required - passing `wpcom.getCart` and `wpcom.setCart` directly gives `TypeError: "this is undefined"`.

**Review by sirbrillig — APPROVED** [review]
Code looks good to me. I haven't tested this last push manually but if you've done so, I'd consider this good to go. Testing before the last two commits worked great.
